sourcegraph/sourcegraph
# Embeddings: support forcing a reindex

This PR adds a `force` flag to the GraphQL API to allow admins to kick off an embeddings job even when there is already an index with that revision. This is helpful in a few cases:
* Fixing prod issues: if we think there's something wrong with the index, admins can make sure `embeddings.incremental` is disabled and manually kick off a reindex.
* Updating embeddings model: this gives a (slightly hacky) way to update the model, by again making sure `embeddings.incremental` is disabled and forcing repos to be reindexed.

The flag isn't surfaced in the UI since it's meant to be used in rare cases.

This PR also adds a fail-safe to the incremental indexing logic: if we can't download the previous repo for whatever reason, we fall back to a full reindex. This will also help with fixing prod issues, because if an index is missing or corrupted for whatever reason, an admin can just kick off a new job and the index will be regenerated.

## Test plan

New unit test for embeddings job scheduler. Manual tests:
* Make sure the new GraphQL parameter works as expected, both when incremental indexing is off and on
* Try deleting an index from blobstore and check that we fall back to a full index
